File: solution/java/day01/src/main/java/food/Food.java
package food;

import java.time.LocalDate;
import java.util.UUID;
import java.util.function.Supplier;

public record Food(LocalDate expirationDate,
                   Boolean approvedForConsumption,
                   UUID inspectorId) {
    public boolean isEdible(Supplier<LocalDate> now) {
        return isFresh(now) &&
                canBeConsumed() &&
                hasBeenInspected();
    }

    private boolean isFresh(Supplier<LocalDate> now) {
        return this.expirationDate.isAfter(now.get());
    }

    private boolean hasBeenInspected() {
        return this.inspectorId != null;
    }

    private Boolean canBeConsumed() {
        return this.approvedForConsumption;
    }
}

File: solution/kotlin/day01/src/main/kotlin/food/Food.kt
package food

import java.time.LocalDate
import java.util.*

data class Food(val expirationDate: LocalDate, val approvedForConsumption: Boolean, val inspectorId: UUID?) {
    fun isEdible(now: () -> LocalDate): Boolean {
        return isFresh(now)
                && canBeConsumed()
                && hasBeenInspected()
    }

    private fun isFresh(now: () -> LocalDate) = expirationDate.isAfter(now())
    private fun canBeConsumed() = approvedForConsumption
    private fun hasBeenInspected() = inspectorId != null
}